Card container

colleenmcginnis opened the pull request:
Closes #13 

This creates a `CardContainer` component and two kinds of cards: `CardWithImage` and `CardTextOnly` (these replace `TutorialCard`). A few questions: 

- Is it 🆒 to keep these as two separate card components or could they be combined in some way? 
- I'm trying to set up the thumbnail image to work with how we're using `AppropriateImage` in the Android, iOS, and Unity docs: 

```js
<AppropriateImage
  backgroundSize="100% auto"
  imageId={this.props.exampleThumbnail}
  background={true}
/>
```

Here's how it is looking so far: 

![image](https://user-images.githubusercontent.com/10479155/42719966-c23a76c0-86d3-11e8-9f3a-4737b4391099.png)

davidtheclark commented on src/components/card-text-only/card-text-only.js:
```diff
@@ -0,0 +1,30 @@
+  render() {
+    const { props } = this;
+    return (
+      <div className="col col--12 mb12">
```
I think that the parent component should determine this element's position in the context of a layout. Looks like that responsibility falls under CardContainer.

It may be that we could even eliminate this wrapper div.

davidtheclark commented on src/components/card-container/card-container.js:
```diff
@@ -0,0 +1,32 @@
+            {props.category.title} ({props.cards.length})
+          </h2>
+        </a>
+        <div className="grid grid--gut12 card-container">{props.cards}</div>
```
The `card-container` doesn't do anything, right?

davidtheclark commented on src/components/card-container/card-container.js:
```diff
@@ -0,0 +1,32 @@
+          href={props.category.path}
+          className="unprose mb12 block color-blue-on-hover"
+        >
+          <h2 className="pt60 txt-l" id={categoryID}>
```
This component shouldn't determine its own top padding: the parent should lay it out in relation to other things.

davidtheclark commented on src/components/card-container/card-container.js:
```diff
@@ -0,0 +1,32 @@
+  category: PropTypes.shape({
+    title: PropTypes.string.isRequired,
+    path: PropTypes.string.isRequired
+  }).isRequired,
```
Not clear to me why `category` is an object, instead of `title` and `path` being props? Seems like it would make a simpler component API to flatten that.

davidtheclark commented on src/components/card-container/card-container.js:
```diff
@@ -0,0 +1,32 @@
+    title: PropTypes.string.isRequired,
+    path: PropTypes.string.isRequired
+  }).isRequired,
+  cards: PropTypes.array.isRequired
```
It's an `arrayOf(PropTypes.node)`, right?

davidtheclark commented on src/components/card-text-only/card-text-only.js:
```diff
@@ -0,0 +1,30 @@
+    return (
+      <div className="col col--12 mb12">
+        <a
+          className="color-gray-dark transition shadow-darken10-on-hover round clip inline-block w-full px12 py12 unprose"
```
Could this be `block` instead of `inline-block w-full`?

davidtheclark commented on src/components/card-text-only/card-text-only.js:
```diff
@@ -0,0 +1,30 @@
+          href={props.path}
+          style={{ verticalAlign: 'top' }}
+        >
+          <div className="px6 pb6">
```
Why is there padding both on the div and the a? Could we remove the div?

davidtheclark commented on src/components/card-with-image/card-with-image.js:
```diff
@@ -0,0 +1,32 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+
+class CardWithImage extends React.Component {
```
I agree with the idea in your comment that these two components could be effectively merged into one.

davidtheclark commented on src/components/card-text-only/card-text-only.js:
```diff
@@ -0,0 +1,30 @@
+        <a
+          className="color-gray-dark transition shadow-darken10-on-hover round clip inline-block w-full px12 py12 unprose"
+          href={props.path}
+          style={{ verticalAlign: 'top' }}
```
What is this `verticalAlign` for?

davidtheclark commented on src/components/card-with-image/card-with-image.js:
```diff
@@ -0,0 +1,32 @@
+  title: PropTypes.string.isRequired,
+  path: PropTypes.string.isRequired,
+  description: PropTypes.string.isRequired,
+  thumbnail: PropTypes.node.isRequired
```
Good move allowing the thumbnail to be passed in from the outside.

davidtheclark commented on src/components/card-container/card-container.js:
```diff
@@ -0,0 +1,32 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+
+class CardContainer extends React.PureComponent {
```
Are cards *with* images ever going to coexist with cards *without* images? Would that work, or would it break the layout?

colleenmcginnis commented:
@davidtheclark I addressed your comments and combined the two card types into one. I also added an `ExamplesPage` component that can be used to make consistent how the `CardContainer`s should be arranged on SDK Examples pages (this makes me feel better about addressing comments like this https://github.com/mapbox/frontend-documentation/pull/17#discussion_r202549845). Do you want to take another look &mdash; or maybe @danswick could take a peek?

davidtheclark commented:
👍 turning over to @danswick

danswick commented:
This looks great to me! I like the change to a general `card` component and a single `card-container` component. This seems ready to go to me. I just made one small change to the sample gifs to ensure our Firefox friends can view them 😄
